### ISSUE TITLE
False positive using constrained TypeVar as return type

### ISSUE THREAD

**Issue by carderne** [microsoft/pyright#4592]
**Describe the bug**
```python
from typing import TypeVar
T = TypeVar("T", str, None)
def foo(val: T) -> T:
    if val is None:
        return None
    return "foo"
```

Produces:
```
Expression of type "None" cannot be assigned to return type "T@foo"
  Type "None" cannot be assigned to type "T@foo"
```
and
```
Expression of type "Literal['foo']" cannot be assigned to return type "T@foo"
  Type "Literal['foo']" cannot be assigned to type "T@foo"
```

**Expected behavior**
This should be fine (mypy doesn't flag it).

**VS Code extension or command-line**
Running in Neovim with neovim-lsp.

**Comment by erictraut** [comment]
This is expected behavior in this case, so I don't consider it a bug. It's a known limitation given the way that pyright handles constrained TypeVars. You can read more about this in [this documentation](https://github.com/microsoft/pyright/blob/main/docs/mypy-comparison.md#constrained-type-variables).

If you want this code to work with pyright without type errors, you'll need to work around this limitation. You could use a `cast` or a `# type: ignore`, for example. You can also return some expression that includes `val`. For example:

```python
def foo(val: T) -> T:
    if val is None:
        return val
    return "foo" or val
```
